### PR TITLE
fix: new url handling approach

### PR DESCRIPTION
**Pull request by inckie**
Use new way for registering url scheme handlers to open internal native screens.

Probably we also need to handle unsupported scheme better in debug and release:
 - Debug (zapp debug, not `__DEV__`): show a screen with message 'scheme is not supported' and button "dismiss". This screen can also list all supported schemes. Dismiss will navigate to home screen.
- Release: just log error that scheme is not recognized and navigate home right away.

Right now I got that and I cant leave this screen without killing the app:
image

### CODE REVIEW

**Comment by inckie** [comment]
@f-roland @Budaa Please review, I'm a bit uncomfortable with having fields inside the function I export, but maybe its ok. Since I have both React screen and special handlers in the same plugin. But maybe its normal approach.
Also please review my note about better error handling.

**Review by Budaa — APPROVED** [review]
LGTM :+1:
Thanks for pointing out the issue with error handling. We're aware of it and I believe is on the TODO list, after project alpha.

**Comment by inckie** [comment]
@Budaa 
This is how it can look like, Sherwin built that awesome screen. Very good for release managers and non js developers.
didomiErrorScreen

**Comment by Budaa** [comment]
@inckie I tought you're talking about general error handling. Yes, I agree, there should be better error handling for this feature and it definitely shouldn't lock you out from the app like that... We will have look.

File: plugins/quick-brick-opta-stats/src/index.js [hunk 1]
```diff
@@ -85,83 +58,11 @@ schemeHandlerHooks["copa_stats"] = async ({ query, url, onFinish }) => {
...
-      // todo: handle errors and fire exit action right away after showing some error message for the user
-    })();
-
-    return () => {
-      logger.info("Unmounted native screen");
-      onDismiss();
-    };
-  }, []);
-  return <View style={styles.container}></View>;
+export default StatScreens = ({ screenData }: Props) => {
```
Comment (f-roland) [inline plugins/quick-brick-opta-stats/src/index.js:61]: what is this `screenData` prop ? it's not being passed to anything here

File: plugins/quick-brick-opta-stats/src/index.js [hunk 2]
```diff
@@ -26,26 +18,7 @@ type Props = {
...
-    </View>
-  ) : (
-    <View style={stylesError.container} onLayout={onDismiss}>
-      <Text style={stylesError.message}>Could not open native screen</Text>
-    </View>
-  );
-}
-
-schemeHandlerHooks["copa_stats"] = async ({ query, url, onFinish }) => {
+useUrlSchemeHandler = async ({ query, url, onFinish }) => {
```
Comment (f-roland) [inline plugins/quick-brick-opta-stats/src/index.js:21]: Have a look here: https://github.com/applicaster/developer-docs-3/pull/80/files

this isn't published yet, but it explains the API to use for the hook.
`onFinish` should be invoked with a callback, calling the `done` function

```
onFinish((done) => {
   // navigation, any other action here...
   done();
});

File: plugins/quick-brick-opta-stats/src/index.js [hunk 3]
```diff
@@ -85,83 +58,11 @@ schemeHandlerHooks["copa_stats"] = async ({ query, url, onFinish }) => {
...
-      logger.info("Unmounted native screen");
-      onDismiss();
-    };
-  }, []);
-  return <View style={styles.container}></View>;
+export default StatScreens = ({ screenData }: Props) => {
+  return <OptaStatsContainer style={styles.container}></OptaStatsContainer>
 };
+
+StatScreens.urlScheme = {
```
Comment (f-roland) [inline plugins/quick-brick-opta-stats/src/index.js:65]: this is fine. in JS everything is an object, even functions. so you can add properties to them

**Review by f-roland — CHANGES_REQUESTED** [review]
adding props to the StatsScreen export is fine

As @Budaa explained we know we need to work on our error screen. This has been in the backlog for a while. We'll look into it.

I thought we already had a mecanism to send home when a scheme couldn't be resolved, but apparently not, at least not for schemes created from plugins (which you're the first one to use)

That, we can add

Please check the API doc I added in the comments to see how to invoke the `onFinish` function. I think this isn't final as the hook isn't doing anything for now ?

File: plugins/quick-brick-opta-stats/src/index.js (hunk 2 above)
Comment (inckie) [reply plugins/quick-brick-opta-stats/src/index.js:15]: added the callback

**Comment by inckie** [comment]
> Please check the API doc I added in the comments to see how to invoke the `onFinish` function. I think this isn't final as the hook isn't doing anything for now ?

Hook is opening a native screen as a new activity. It's was the desired approach there and in Didomi plugin: call a native method and wait for the result without any additional screens or navigation in the QB app.

**Review by f-roland — COMMENTED** [review]
I see.
Did we test what happens when a user wants to leave these screens ? are you getting back to the home screen ? an empty screen ? 

With the approach you describe, basically it's like you are opening a new app, so I'm not sure what the user experience will be when the user is pressing the back button from this screen

**Comment by inckie** [comment]
@f-roland app loads to the home screen first in any case, and only then url processing happens, so when you navigate back, you get to the original state.

**Comment by f-roland** [comment]
alright, looks good then

**Review by f-roland — APPROVED** [review]

**Review by kononenkoAnton — APPROVED** [review]

**Pull request merged**